synapse/pipeline.yml

```diff
@@ -527,3 +527,34 @@ steps:
 #
 #  - label: Trigger webhook
 #    command: "curl -k https://coveralls.io/webhook?repo_token=$COVERALLS_REPO_TOKEN -d \"payload[build_num]=$BUILDKITE_BUILD_NUMBER&payload[status]=done\""
+
+  ################################################################################
+  #
+  # Complement Test Suite
+  #
+  ################################################################################
+
+  - command:
+      - "docker build -t matrixdotorg/synapse:latest -f docker/Dockerfile ."
+      - "wget -N https://github.com/matrix-org/complement/archive/master.tar.gz && tar -xzf master.tar.gz"
+      - "docker build -t complement-synapse -f complement-master/dockerfiles/Synapse.Dockerfile complement-master/dockerfiles"
+      - "cd complement-master && COMPLEMENT_BASE_IMAGE=complement-synapse:latest go test -v ./tests"
+    label: "\U0001F9EA Complement"
+    agents:
+      queue: "medium"
+    # Continue even on failure for now
+    soft_fail:
+      - exit_status: 1
+    plugins:
+      - docker#v3.5.0:
+          # The dockerfile for this image is at https://github.com/matrix-org/complement/blob/master/dockerfiles/ComplementCIBuildkite.Dockerfile.
+          image: "matrixdotorg/complement:latest"
+          mount-buildkite-agent: false
+          # Complement needs to know if it is running under CI
+          environment:
+           - "CI=true"
+          publish: [ "8448:8448" ]
+          # Complement uses Docker so pass through the docker socket. This means Complement shares
+          # the hosts Docker.
+          volumes:
+            - "/var/run/docker.sock:/var/run/docker.sock"
```

Comment:
hmm - I don't really know enough about docker to reason whether this is safe or not. Would it not let you start inspecting other running containers from other buildkite jobs and/or killing them off?

Reply:
That's correct, and I'm mostly leaning on the fact that Dendrite does the same thing and a security discussion has already happened [here](https://github.com/matrix-org/pipelines/pull/87) :D

Reply:
yes it would. We're trusting the complement master branch and docker image not to do nasty things.

It looks like a difference between https://github.com/matrix-org/pipelines/pull/87 and here is that we are running a `docker build` for the synapse branch under test within the privileged container. I suspect that could be used to construct a container escape, which is problematic.

I don't know where the dendrite docker imaged used in #87 comes from.

Reply:
Right, a community member reminded me that this was added in https://github.com/matrix-org/pipelines/pull/92, and is built from the Dendrite repo. Is this a concern?